### PR TITLE
[VL]  parquet file metadata columns support in velox

### PR DESCRIPTION
**Pull request by gaoyangxiaozhu**
## What changes were proposed in this pull request?

Support file metadata column velox native access, it is requirement of Delta native support for Delta Microsoft Team asked.

Fixes: https://github.com/oap-project/gluten/issues/2618

## How was this patch tested?

Via UT and manually tests

image

dependency PR in velox is https://github.com/facebookincubator/velox/pull/8800 and is merged 

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by gaoyangxiaozhu** [comment]
@zhouyuan and @FelixYBW could you help check why centos7-test fail and give some input ? Meanwhile could you help have a draft review to see if current implement for metadata column native support good to you , if the implement ok for you guys , i will sync with meta velox guys for velox part PR review.

**Comment by yma11** [comment]
@gaoyangxiaozhu Thanks for providing this support. The implementation seems okay. Please do a rebase and go ahead with corresponding support in Velox first.

**Comment by gaoyangxiaozhu** [comment]
will re-start the work as the Velox part is ready

**Comment by zhli1142015** [comment]
THe CI faulire looks related to your change @gaoyangxiaozhu, please check, thanks.

File: cpp/velox/substrait/SubstraitParser.cc [hunk 1]
```diff
@@ -104,31 +104,41 @@ std::vector<TypePtr> SubstraitParser::parseNamedStruct(const ::substrait::NamedS
...
-std::vector<bool> SubstraitParser::parsePartitionColumns(const ::substrait::NamedStruct& namedStruct) {
+void SubstraitParser::parsePartitionAndMetadataColumns(
+    const ::substrait::NamedStruct& namedStruct,
+    std::vector<bool>& isPartitionColumns,
+    std::vector<bool>& isMetadataColumns) {
   const auto& columnsTypes = namedStruct.column_types();
-  std::vector<bool> isPartitionColumns;
   if (columnsTypes.size() == 0) {
-    // Regard all columns as non-partitioned columns.
+    // Regard all columns as regular columns.
```
Comment (yma11) [inline cpp/velox/substrait/SubstraitParser.cc:113]: If `columnsTypes` has 0 as size, does it mean that there is no columns in this table? if so, maybe we should do a check `columnsTypes.size() > 0` instead?

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenFileMetadataStructSuite.scala [hunk 2]
```diff
@@ -16,6 +16,156 @@
...
+      assert(fileScan.size == 1)
+      if (BackendTestUtils.isVeloxBackendLoaded()) {
+        assert(fileScan(0).nodeNamePrefix == "NativeFile")
+      } else {
+        assert(fileScan(0).nodeNamePrefix == "File")
+      }
+      var filterExecs = filterDF.queryExecution.executedPlan.collect {
+        case filter: FilterExecTransformerBase => filter
+      }
+      assert(filterExecs.size == 1)
```
Comment (zhouyuan) [inline gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenFileMetadataStructSuite.scala:149]: the goal is to check the scan is executed natively. can we use similar approach: 
https://github.com/apache/incubator-gluten/blob/main/backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala#L467

**Comment by zhouyuan** [comment]
Quick note: This patch adds the native impl for parquet metadata read support which is a requirement from delta lake connector. The general idea is to add a metadata list param for filescan/batchscan which will be filled in Velox native scan.

Thanks,
-yuan

File: shims/spark34/src/main/scala/io/glutenproject/sql/shims/spark34/Spark34Shims.scala [hunk 3]
```diff
@@ -153,6 +157,34 @@ class Spark34Shims extends SparkShims {
...
+          metadataColumn.put(FileFormat.FILE_MODIFICATION_TIME, fileModifyTime)
+        case FileFormat.FILE_BLOCK_START =>
+          metadataColumn.put(FileFormat.FILE_BLOCK_START, file.start.toString)
+        case FileFormat.FILE_BLOCK_LENGTH =>
+          metadataColumn.put(FileFormat.FILE_BLOCK_LENGTH, file.length.toString)
+        case _ =>
+      }
+    }
+
+    // TODO row_index metadata support
```
Comment (yma11) [inline shims/spark34/src/main/scala/io/glutenproject/sql/shims/spark34/Spark34Shims.scala:184]: `TODO: `

File: cpp/velox/substrait/SubstraitParser.cc (hunk 1 above)
Comment (gaoyangxiaozhu) [reply cpp/velox/substrait/SubstraitParser.cc:113]: for read node it should be true, but not sure for write node if `namedStruct.names()` also is 0 size.

Comment (yma11) [reply cpp/velox/substrait/SubstraitParser.cc:113]: @JkSelf will it happen for parquet write?

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenFileMetadataStructSuite.scala (hunk 2 above)
Comment (yma11) [reply gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenFileMetadataStructSuite.scala:149]: +1

File: cpp/velox/substrait/SubstraitParser.cc (hunk 1 above)
Comment (gaoyangxiaozhu) [reply cpp/velox/substrait/SubstraitParser.cc:113]: so if is true for write then we only process under `columnsTypes.size() > 0` ?

Comment (yma11) [reply cpp/velox/substrait/SubstraitParser.cc:113]: I think so but we'd better confirm first.

**Comment by gaoyangxiaozhu** [comment]
various build issue , can you help re-trigger @yma11 / @zhouyuan / @zhli1142015 
image

**Comment by gaoyangxiaozhu** [comment]
can we re-trigger again for failed job which all due to below

image

**Review by yma11 — APPROVED** [review]

**Comment by zhouyuan** [comment]
@gaoyangxiaozhu This is in good state to me. Will try with internal delta lake jenkins job 
CC @zzcclp as this will also change the API for CK backend

**Comment by yma11** [comment]
@gaoyangxiaozhu seems there is some conflicts and you need a rebase. Thanks.

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu seems there is some conflicts and you need a rebase. Thanks.

done

**Review by zhouyuan — APPROVED** [review]
👍

**Pull request merged**